File: code_100/java/com/interview/array/CommonThreeSortedArray.java
package com.interview.array;

import java.util.ArrayList;
import java.util.List;

/**
 * Date 01/01/2016
 * @author Tushar Roy
 *
 * Given 3 sorted array find common elements in these 3 sorted array.
 *
 * Time complexity is O(m + n + k)
 *
 * http://www.geeksforgeeks.org/find-common-elements-three-sorted-arrays/
 */
public class CommonThreeSortedArray {

    public List<Integer> commonElements(int input1[], int input2[], int input3[]) {
        int i = 0;
        int j = 0;
        int k = 0;
        List<Integer> result = new ArrayList<>();
        while (i < input1.length && j < input2.length && k < input3.length) {
            if (input1[i] == input2[j] && input2[j] == input3[k]) {
                result.add(input1[i]);
                i++;
                j++;
                k++;
            } else if (input1[i] < input2[j]) {
                i++;
            } else if (input2[j] < input3[k]) {
                j++;
            } else {
                k++;
            }
        }
        return result;
    }

    public static void main(String args[]) {
        int input1[] = {1, 5, 10, 20, 40, 80};
        int input2[] = {6, 7, 20, 80, 100};
        int input3[] = {3, 4, 15, 20, 30, 70, 80, 120};

        CommonThreeSortedArray cts = new CommonThreeSortedArray();
        List<Integer> result = cts.commonElements(input1, input2, input3);
        result.forEach(i -> System.out.print(i + " "));
    }
}

File: code_100/python/array/commonthreesortedarray.py
# http://www.geeksforgeeks.org/find-common-elements-three-sorted-arrays/

def common_elements(input1, input2, input3):
    result = []
    i = 0
    j = 0
    k = 0
    while i < len(input1) and j < len(input2) and k < len(input3):
        if input1[i] == input2[j] and input2[j] == input3[k]:
            result.append(input1[i])
            i = i + 1
            j = j + 1
            k = k + 1
        elif input1[i] < input2[j]:
            i = i + 1
        elif input2[j] < input3[k]:
            j = j + 1
        else:
            k = k + 1
    return result

if __name__ == '__main__':
    input1 = [1, 5, 10, 20, 40, 80]
    input2 = [6, 7, 20, 80, 100]
    input3 = [3, 4, 15, 20, 30, 70, 80, 120]

    print(common_elements(input1, input2, input3))
